README.md
# Dash:next

Experimental UI enhancements for Zyte [dashboard](https://app.zyte.com/) via browser extension.

## Purpose

The extension serves as a playground to try out some useful UI elements and features.
The successful ones might end up being incorporated into Dash.

## List of Enhancements

1. `job-stats`

    Adds chronological stats data viz to the main job page (i.e. `https://app.zyte.com/p/*/*/*`).
    Every stats field is displayed as a time series of relative changes for the last 7 jobs.
    The obtained data viz can be exported as SVG/PNG file.

## Installation

Currently the extension is alpha, so:

- it's not branded (no icons, "About" pages, etc.);
- it's not published to Chrome Web Store;
- it's not ported to Firefox.

You could try it out in [Chrome's Developer mode](https://developer.chrome.com/extensions/getstarted#manifest):

1. clone it `git clone git@github.com:alexander-matsievsky/dash-next.git /tmp/dash-next`;
1. open the Extension Management page by navigating to `chrome://extensions`;
1. enable Developer Mode by clicking the toggle switch next to **Developer mode**;
1. click the **LOAD UNPACKED** button and select `/tmp/dash-next`.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.3](https://github.com/alexander-matsievsky/dash-next/compare/v0.0.2...v0.0.3) (2021-06-29)


### Features

* switch from `scrapinghub.com` to `zyte.com` after rebranding ([5fd6597](https://github.com/alexander-matsievsky/dash-next/commit/5fd65978e369637aa6bc68b0078a56f4766bae39))

### [0.0.2](https://github.com/alexander-matsievsky/dash-next/compare/v0.0.1...v0.0.2) (2020-03-31)


### Features

* **periodic-jobs:** new enhancement ([f857071](https://github.com/alexander-matsievsky/dash-next/commit/f85707177e0735065bce38f0eba5a8659485b6a2))

### 0.0.1 (2019-12-28)


### Features

* **job-stats:** new enhancement ([dde7cdc](https://github.com/alexander-matsievsky/dash-next/commit/dde7cdc980544208571b48ce1e3b961ac8c5e216))


content_scripts/router.js
const routes = [
  {
    // eslint-disable-next-line no-undef
    handler: JobStats.inject,
    pattern: new RegExp('^/p/(?<projectId>\\d+)/(?<spiderId>\\d+)/(?<jobId>\\d+)$')
  },
  {
    // eslint-disable-next-line no-undef
    handler: PeriodicJobs.inject,
    pattern: new RegExp('^/p/(?<projectId>\\d+)/periodic-jobs$')
  }
]

function dispatchRoute ({ type, url }) {
  url = new URL(url)
  if (type !== 'router' || url.origin !== 'https://app.zyte.com') return
  for (const { handler, pattern } of routes) {
    const match = url.pathname.match(pattern)
    if (match) handler(match.groups)
  }
}

[
  'https://cdn.jsdelivr.net/npm/vega@5.10.0',
  'https://cdn.jsdelivr.net/npm/vega-lite@4.8.1',
  'https://cdn.jsdelivr.net/npm/vega-embed@6.5.1'
].reduce(
  async (loaded, src) => {
    await loaded
    return new Promise((resolve, reject) => {
      const script = Object.assign(document.createElement('script'), { src })
      script.addEventListener('load', resolve)
      script.addEventListener('error', reject)
      document.head.appendChild(script)
    })
  },
  Promise.resolve()
).then(() => {
  dispatchRoute({ type: 'router', url: document.location.href })
  chrome.runtime.onMessage.addListener(dispatchRoute)
})


content_scripts/job-stats.js
// eslint-disable-next-line no-unused-vars
async function jobStats ({ jobId, projectId, spiderId }, retry = 0) {
  if (retry >= 3) {
    return console.error(`missed #job-stats ${retry} times`)
  }
  if (!document.getElementById('job-stats')) {
    return setTimeout(() => jobStats({ jobId, projectId, spiderId }, retry + 1), 1e3)
  }

  const jobKey = `${projectId}/${spiderId}/${jobId}`
  const JOB_COUNT = 7
  const { apikey } = await (await fetch('/api/v2/config')).json()
  const job = await (await fetch(`https://storage.scrapinghub.com/jobs/${jobKey}?apikey=${apikey}`)).json()
  const list = await (await fetch(`/api/jobs/list.json?project=${projectId}&spider=${job.spider}`)).json()
  if (!list.jobs.find(_ => _.id === jobKey)) list.jobs.push({ id: jobKey })
  list.jobs.sort((a, b) => +a.id.split('/')[2] - +b.id.split('/')[2])
  const jobI = list.jobs.findIndex(_ => _.id === jobKey)
  const jobs = await Promise.all(
    list.jobs.slice(Math.max(0, jobI + 1 - JOB_COUNT), jobI + 1).map(async ({ id }) => {
      const job = await (await fetch(`https://storage.scrapinghub.com/jobs/${id}?apikey=${apikey}`)).json()
      const stats = await (await fetch(`https://storage.scrapinghub.com/items/${id}/stats?apikey=${apikey}&format=json`)).json()
      const runtimeHours = ((job.finished_time || Date.now()) - (job.running_time || Date.now())) / 1000 / 60 / 60
      return {
        ...job,
        ...(job.scrapystats || {}),
        ...(stats.counts || {}),
        fieldstats: Object.keys(stats.counts || {}),
        job_key: id,
        runtime_hours: runtimeHours,
        scrapystats: Object.keys(job.scrapystats || {})
      }
    })
  )

  const statsGroups = [
    [
      'FIELDS',
      Array.from(new Set(jobs.map(_ => _.fieldstats).flat()))
        .filter(_ => !/\/|^_type$/.test(_))
        .sort()
    ],
    [
      'ITEMS',
      Array.from(new Set(jobs.map(_ => _.scrapystats).flat()))
        .filter(_ => /^item_/.test(_))
        .concat('runtime_hours')
        .sort()
    ],
    [
      'LOGS',
      Array.from(new Set(jobs.map(_ => _.scrapystats).flat()))
        .filter(_ => /^log_/.test(_))
        .sort()
    ],
    [
      'NETWORK',
      Array.from(new Set(jobs.map(_ => _.scrapystats).flat()))
        .filter(_ => /^(crawlera|downloader|dupefilter|httperror)/.test(_))
        .sort()
    ]
  ]
  statsGroups.push([
    'OTHER STATS',
    Array.from(
      function * () {
        const fieldstats = new Set(jobs.map(_ => _.fieldstats).flat())
        const groupedstats = new Set(statsGroups.map(_ => _[1]).flat())
        const scrapystats = new Set(jobs.map(_ => _.scrapystats).flat())
        for (const field of scrapystats) {
          if (groupedstats.has(field) || fieldstats.has(field)) continue
          if (['finish_reason', 'finish_time', 'pending_time', 'start_time'].includes(field)) continue
          yield field
        }
      }.call()
    ).sort()
  ])
  const datasets = statsGroups.reduce(
    (datasets, [title, fields]) => Object.assign(datasets, {
      [title]: jobs.flatMap((job, i, jobs) => {
        const curr = job
        const prev = jobs[i - 1] || {}
        const finishedTime = new Date(job.finished_time || Date.now()).toISOString()
        return fields.map(field => {
          const countCurr = curr[field] || 0
          const countPrev = prev[field] || 0
          const countChange = Object.keys(prev).length ? countCurr - countPrev : 0
          const countChangePct = +(countChange / countPrev * 100 || 0).toFixed(2)
          return {
            'finished on': finishedTime,
            'job href': `https://app.scrapinghub.com/p/${job.job_key}`,
            'job key': job.job_key,
            'job number': +job.job_key.split('/')[2],
            'Δcount %': Math.max(-100, Math.min(countChangePct, 100)),
            count: countCurr,
            field: field,
            Δcount: countChange
          }
        })
      })
    }),
    {}
  )

  const { width } = document.getElementById('job-stats').getBoundingClientRect()
  const columns = (width / 225 | 0) - 1
  const vega = {
    $schema: 'https://vega.github.io/schema/vega-lite/v4.0.0-beta.10.json',
    background: 'white',
    config: { mark: { tooltip: null } },
    datasets: datasets,
    title: `https://app.scrapinghub.com/p/${projectId}/${spiderId}/${jobId} ${job.spider}`,
    vconcat: Object.keys(datasets).map(title => ({
      columns: columns,
      data: { name: title },
      facet: { field: 'field', type: 'nominal', header: { labelLimit: 175 } },
      resolve: { scale: { y: 'independent' } },
      spec: {
        layer: [
          {
            encoding: {
              x: { field: 'job number', type: 'ordinal' },
              y: {
                axis: { maxExtent: 75, minExtent: 75 },
                field: 'count',
                scale: { zero: false },
                title: null,
                type: 'quantitative'
              }
            },
            height: 50,
            mark: {
              interpolate: 'basis',
              stroke: 'lightgray',
              type: 'line'
            },
            width: 100
          },
          {
            encoding: {
              color: {
                field: 'Δcount %',
                scale: { scheme: 'redblue', type: 'symlog' },
                type: 'quantitative'
              },
              href: { field: 'job href', type: 'nominal' },
              tooltip: [
                { field: 'field', type: 'nominal' },
                { field: 'count', type: 'quantitative' },
                { field: 'Δcount %', type: 'quantitative' },
                { field: 'Δcount', type: 'quantitative' },
                { field: 'job key', type: 'nominal' },
                { field: 'finished on', type: 'temporal' }
              ],
              x: { field: 'job number', type: 'ordinal' },
              y: {
                axis: { maxExtent: 75, minExtent: 75 },
                field: 'count',
                scale: { zero: false },
                title: null,
                type: 'quantitative'
              }
            },
            height: 50,
            mark: {
              filled: true,
              size: 75,
              stroke: 'gray',
              type: 'point'
            },
            width: 100
          }
        ]
      },
      title: title
    }))
  }
  console.debug(vega)
  document.getElementById('job-stats').after(
    Object.assign(document.createElement('div'), {
      className: 'panel',
      id: 'job-stats-next',
      innerHTML: '<div></div>',
      style: 'margin-top:34px;'
    })
  )
  document.getElementById('job-stats-next').appendChild(
    Object.assign(document.createElement('script'), {
      innerHTML: `vegaEmbed('#job-stats-next>div',${(JSON.stringify(vega))})`
    })
  )
}


content_scripts/PeriodicJobs.js
class PeriodicJobs extends HTMLElement {
  constructor () {
    super()
    this.projectId = this.getAttribute('project-id')
    this.schedule = this.fetchSchedule()
  }

  static async inject ({ projectId }) {
    const selectors = '.dash-page>div:first-child'
    const periodicJobs = new window.DOMParser().parseFromString(
      `<dash-next-periodic-jobs project-id="${projectId}">`,
      'text/html'
    ).querySelector('dash-next-periodic-jobs')
    await new Promise(resolve => {
      const interval = window.setInterval(() => {
        if (document.querySelector(selectors)) {
          resolve(window.clearInterval(interval))
        }
      }, 500)
    })
    document.querySelector(selectors).after(periodicJobs)
  }

  connectedCallback () {
    window.addEventListener('resize', () => this.render())
    this.render()
  }

  async fetchSchedule () {
    const calendar = Array(31)
      .fill(new Date(Date.now()))
      .map((now, i) => new Date(now.setUTCDate(i + 1)))
      .filter(function (date) {
        return date.getUTCMonth() === this.getUTCMonth() &&
          date.getUTCFullYear() === this.getUTCFullYear()
      }, new Date(Date.now()))
      .map((date, _, dates) => ({
        date: date,
        day: date.getUTCDay() || 7,
        dayofmonth: date.getUTCDate(),
        month: date.getUTCMonth() + 1,
        weekofmonth: (((dates[0].getUTCDay() || 7) - 2 + date.getUTCDate()) / 7 | 0) + 1
      }))
    let { results: periodicJobs } = await (
      await fetch(`/api/v2/projects/${this.projectId}/periodicjobs`)
    ).json()
    periodicJobs = periodicJobs
      .filter(periodicJob => !periodicJob.disabled)
      .flatMap(periodicJob =>
        periodicJob.spiders.map(spider => {
          periodicJob.spider_name = spider.name
          delete periodicJob.spider
          delete periodicJob.spiders
          return periodicJob
        })
      )
    const schedule = calendar.flatMap(date => {
      const schedule = periodicJobs
        .filter(periodicJob =>
          (periodicJob.month === '*' || +periodicJob.month === date.month) &&
          (periodicJob.day === '*' || +periodicJob.day === date.day) &&
          (periodicJob.dayofmonth === '*' || +periodicJob.dayofmonth === date.dayofmonth)
        )
        .map(periodicJob => ({ ...periodicJob, ...date, periodicjob: 1 }))
      return schedule.length ? schedule : { ...date, periodicjob: 0 }
    }).sort((a, b) =>
      a.dayofmonth - b.dayofmonth || a.spider_name.localeCompare(b.spider_name)
    )
    let spiderNamesMap = schedule.reduce((spiderNamesMap, schedule) => {
      const spiderNames = spiderNamesMap.get(schedule.dayofmonth) || new Set()
      spiderNames.add(schedule.spider_name)
      spiderNamesMap.set(schedule.dayofmonth, spiderNames)
      return spiderNamesMap
    }, new Map())
    spiderNamesMap = Object.fromEntries(
      [...spiderNamesMap.entries()].map(([k, v]) => [k, Array.from(v).sort().join('\n')])
    )
    return schedule.map(schedule =>
      Object.assign(schedule, { spider_names: spiderNamesMap[schedule.dayofmonth] })
    )
  }

  async render () {
    const schedule = await this.schedule
    const title = new Date(Date.now()).toLocaleString(
      window.navigator.language, { month: 'long', year: 'numeric' }
    )
    const vega = {
      $schema: 'https://vega.github.io/schema/vega-lite/v4.0.2.json',
      config: {
        legend: {
          orient: 'left'
        },
        view: {
          continuousHeight: 360,
          continuousWidth: 400,
          stroke: 'transparent'
        }
      },
      data: {
        name: 'schedule'
      },
      datasets: {
        schedule: schedule
      },
      layer: [
        {
          encoding: {
            color: {
              aggregate: 'sum',
              field: 'periodicjob',
              scale: {
                scheme: 'lightgreyred'
              },
              title: 'Jobs',
              type: 'quantitative'
            },
            tooltip: [
              {
                aggregate: 'sum',
                field: 'periodicjob',
                title: 'Jobs',
                type: 'quantitative'
              },
              {
                field: 'spider_names',
                title: 'Spiders',
                type: 'ordinal'
              }
            ],
            x: {
              axis: null,
              field: 'day',
              type: 'ordinal'
            },
            y: {
              axis: null,
              field: 'weekofmonth',
              type: 'ordinal'
            }
          },
          height: 360,
          mark: {
            cornerRadius: 5,
            height: 55,
            type: 'rect',
            width: 55
          },
          title: title,
          width: 420
        },
        {
          encoding: {
            text: {
              field: 'dayofmonth',
              type: 'ordinal'
            },
            tooltip: [
              {
                aggregate: 'sum',
                field: 'periodicjob',
                title: 'Jobs',
                type: 'quantitative'
              },
              {
                field: 'spider_names',
                title: 'Spiders',
                type: 'ordinal'
              }
            ],
            x: {
              axis: null,
              field: 'day',
              type: 'ordinal'
            },
            y: {
              axis: null,
              field: 'weekofmonth',
              type: 'ordinal'
            }
          },
          height: 360,
          mark: {
            color: 'white',
            fontSize: 30,
            type: 'text'
          },
          width: 420
        }
      ]
    }
    this.innerHTML = `
      <div class="panel" id="dash-next-periodic-jobs" style="padding-top:12px;">
        <div></div>
      </div>
    `
    this.appendChild(
      Object.assign(document.createElement('script'), {
        innerHTML: `vegaEmbed('#dash-next-periodic-jobs>div',${(JSON.stringify(vega))})`
      })
    )
  }
}

window.customElements.define('dash-next-periodic-jobs', PeriodicJobs)


content_scripts/JobStats.js
class JobStats extends HTMLElement {
  constructor () {
    super()
    this.JOB_COUNT = 7
    this.jobId = this.getAttribute('job-id')
    this.projectId = this.getAttribute('project-id')
    this.spiderId = this.getAttribute('spider-id')
    this.jobKey = `${this.projectId}/${this.spiderId}/${this.jobId}`
    this.spiderName = '<spider>'
    this.datasets = this.fetchDatasets()
  }

  static async inject ({ jobId, projectId, spiderId }) {
    const jobStats = new window.DOMParser().parseFromString(
      `
      <dash-next-job-stats
        job-id="${jobId}"
        project-id="${projectId}"
        spider-id="${spiderId}"
      >
      `,
      'text/html'
    ).querySelector('dash-next-job-stats')
    await new Promise(resolve => {
      const interval = window.setInterval(() => {
        if (document.getElementById('job-stats')) {
          resolve(window.clearInterval(interval))
        }
      }, 500)
    })
    document.getElementById('job-stats').after(jobStats)
  }

  connectedCallback () {
    window.addEventListener('resize', () => this.render())
    this.render()
  }

  async fetchDatasets () {
    const { apikey } = await (await fetch('/api/v2/config')).json()
    const job = await (await fetch(
      `https://storage.scrapinghub.com/jobs/${this.jobKey}?apikey=${apikey}`
    )).json()
    const list = await (await fetch(
      `/api/jobs/list.json?project=${this.projectId}&spider=${job.spider}`
    )).json()
    if (!list.jobs.find(({ id }) => id === this.jobKey)) {
      list.jobs.push({ id: this.jobKey })
    }
    list.jobs.sort((a, b) => +a.id.split('/')[2] - +b.id.split('/')[2])

    this.spiderName = job.spider
    const jobI = list.jobs.findIndex(_ => _.id === this.jobKey)
    const jobs = await Promise.all(
      list.jobs.slice(Math.max(0, jobI + 1 - this.JOB_COUNT), jobI + 1)
        .map(async ({ id }) => {
          const job = await (await fetch(
              `https://storage.scrapinghub.com/jobs/${id}?apikey=${apikey}`)
          ).json()
          const stats = await (await fetch(
              `https://storage.scrapinghub.com/items/${id}/stats?apikey=${apikey}&format=json`)
          ).json()
          const runtimeHours = (
            (job.finished_time || Date.now()) - (job.running_time || Date.now())
          ) / 1000 / 60 / 60
          return {
            ...job,
            ...(job.scrapystats || {}),
            ...(stats.counts || {}),
            fieldstats: Object.keys(stats.counts || {}),
            job_key: id,
            runtime_hours: runtimeHours,
            scrapystats: Object.keys(job.scrapystats || {})
          }
        })
    )
    const statsGroups = [
      [
        'FIELDS',
        Array.from(new Set(jobs.map(_ => _.fieldstats).flat()))
          .filter(_ => !/\/|^_type$/.test(_))
          .sort()
      ],
      [
        'ITEMS',
        Array.from(new Set(jobs.map(_ => _.scrapystats).flat()))
          .filter(_ => /^item_/.test(_))
          .concat('runtime_hours')
          .sort()
      ],
      [
        'LOGS',
        Array.from(new Set(jobs.map(_ => _.scrapystats).flat()))
          .filter(_ => /^log_/.test(_))
          .sort()
      ],
      [
        'NETWORK',
        Array.from(new Set(jobs.map(_ => _.scrapystats).flat()))
          .filter(_ => /^(crawlera|downloader|dupefilter|httperror)/.test(_))
          .sort()
      ]
    ]
    statsGroups.push([
      'OTHER STATS',
      Array.from(
        function * () {
          const fieldstats = new Set(jobs.map(_ => _.fieldstats).flat())
          const groupedstats = new Set(statsGroups.map(_ => _[1]).flat())
          const scrapystats = new Set(jobs.map(_ => _.scrapystats).flat())
          for (const field of scrapystats) {
            if (groupedstats.has(field) || fieldstats.has(field)) continue
            if ([
              'finish_reason',
              'finish_time',
              'pending_time',
              'start_time'
            ].includes(field)) continue
            yield field
          }
        }.call()
      ).sort()
    ])
    return statsGroups.reduce(
      (datasets, [title, fields]) => Object.assign(datasets, {
        [title]: jobs.flatMap((job, i, jobs) => {
          const curr = job
          const prev = jobs[i - 1] || {}
          const finishedTime = new Date(job.finished_time || Date.now()).toISOString()
          return fields.map(field => {
            const countCurr = curr[field] || 0
            const countPrev = prev[field] || 0
            const countChange = Object.keys(prev).length ? countCurr - countPrev : 0
            const countChangePct = +(countChange / countPrev * 100 || 0).toFixed(2)
            return {
              'finished on': finishedTime,
              'job href': `https://app.zyte.com/p/${job.job_key}`,
              'job key': job.job_key,
              'job number': +job.job_key.split('/')[2],
              'Δcount %': Math.max(-100, Math.min(countChangePct, 100)),
              count: countCurr,
              field: field,
              Δcount: countChange
            }
          })
        })
      }),
      {}
    )
  }

  async render () {
    const { width } = document.getElementById('job-stats').getBoundingClientRect()
    const columns = (width / 225 | 0) - 1
    const datasets = await this.datasets
    const vega = {
      $schema: 'https://vega.github.io/schema/vega-lite/v4.8.1.json',
      background: 'white',
      config: { mark: { tooltip: null } },
      datasets: datasets,
      title: `https://app.zyte.com/p/${this.jobKey} ${this.spiderName}`,
      vconcat: Object.keys(datasets).map(title => ({
        columns: columns,
        data: { name: title },
        facet: { field: 'field', type: 'nominal', header: { labelLimit: 175 } },
        resolve: { scale: { y: 'independent' } },
        spec: {
          layer: [
            {
              encoding: {
                x: { field: 'job number', type: 'ordinal' },
                y: {
                  axis: { maxExtent: 75, minExtent: 75 },
                  field: 'count',
                  scale: { zero: false },
                  title: null,
                  type: 'quantitative'
                }
              },
              height: 50,
              mark: {
                interpolate: 'basis',
                stroke: 'lightgray',
                type: 'line'
              },
              width: 100
            },
            {
              encoding: {
                color: {
                  field: 'Δcount %',
                  scale: { domain: [-100, 100], scheme: 'redblue', type: 'symlog' },
                  type: 'quantitative'
                },
                href: { field: 'job href', type: 'nominal' },
                tooltip: [
                  { field: 'field', type: 'nominal' },
                  { field: 'count', type: 'quantitative' },
                  { field: 'Δcount %', type: 'quantitative' },
                  { field: 'Δcount', type: 'quantitative' },
                  { field: 'job key', type: 'nominal' },
                  { field: 'finished on', type: 'temporal' }
                ],
                x: { field: 'job number', type: 'ordinal' },
                y: {
                  axis: { maxExtent: 75, minExtent: 75 },
                  field: 'count',
                  scale: { zero: false },
                  title: null,
                  type: 'quantitative'
                }
              },
              height: 50,
              mark: {
                filled: true,
                size: 75,
                stroke: 'gray',
                type: 'point'
              },
              width: 100
            }
          ]
        },
        title: title
      }))
    }
    this.innerHTML = `
      <div class="panel" id="dash-next-job-stats" style="margin-top:34px;padding-top:12px;">
        <div></div>
      </div>
    `
    this.appendChild(
      Object.assign(document.createElement('script'), {
        innerHTML: `vegaEmbed('#dash-next-job-stats>div',${(JSON.stringify(vega))})`
      })
    )
  }
}

window.customElements.define('dash-next-job-stats', JobStats)
